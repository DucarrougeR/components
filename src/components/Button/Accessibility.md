# Accessibility



## Best Practices
- Buttons **do not** open a new page, use links instead
- Buttons can open dialogs and menus, but should use the proper ARIA properties

## Labeling Expectations
- Use clear and descriptive labels. By default this is the text inside the `<Button>` component.
- Ambiguous labels like "click here", "run", "save" are ok for sighted users but can lack context for users using assistive technology. Use the `aria-label` attribute to provide an accessible label for all `<Button>` components with potentially ambiguous labels.
- Icon only buttons must provide a descriptive label using the `aria-label` attribute.

## Focus Expectations

- Buttons with `:focus` should have an outline and/or a visible change to show it has focus/

## Keyboard Expectations

- Buttons with focus should be activated with user hits the `ENTER` and `SPACEBAR` keys
- If button has focus, `TAB` and `SHIFT+TAB` should move focus to next focusable element.

## Aria Properties

### `aria-label`
Use this to provide an accessible label for `<Button>` components. Good for buttons with ambiguous labels and icon buttons

For example `<Button aria-label="Save Dashboard">Save</Button>` would be announced as "Save Dashboard Button" by a screen reader instead of just "Save Button"

### `aria-haspopup`
Use this to indicate that when the buttons is pressed a popup is triggered. A popup can be a dialog or menu.

For example clicking an icon buttons to reveal a set of options in a context menu you do the following

```
<IconGearButton aria-haspopup="true">
```

If a button opens a modal dialog window you would the following

```
<Button aria-haspopup="dialog">New Schedule</Button>
```


