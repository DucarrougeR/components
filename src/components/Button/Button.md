## Variants

The `<Button />` component accepts a `variant` property which allows you to select the proper button style for the context and action the button is for.

### Primary

Use a primary button for the most frequently used action or most important action on a page. Primary buttons should only appear once on a page. They are sometimes accompanied by a secondary button.

```js
 <Button>Primary Button</Button>
```
##### **Primary states**

```js noeditor
<table>
  <tr>
    <td><Button>Primary Idle</Button></td>
    <td><Button className="hover">Primary Hover</Button></td>
    <td><Button className="active">Primary Active</Button></td>
     <td><Button disabled>Disabled Active</Button></td>
  </tr>
</table>

```

<div class="doc-section-divider"></div>

### Outline
Use an outline button alongside a primary button to provide alternative actions on a page. Outline buttons should only appear alongside primary buttons for secondary actions. Use no more than two secondary buttons on a page.

```js

<List>
  <ListItem>
    <Button variant="outline">Light Idle</Button>
  </ListItem>
  <ListItem>
    <Button variant="outline" state="hover">Light Hover</Button>
  </ListItem>
  <ListItem>
     <Button variant="outline" state="active">Light Active</Button>
  </ListItem>
</List>
```

<div class="doc-section-divider"></div>

### Transparent

Use a transparent button as a tertiary action on a screen, they are often used as a Cancel button on a form. Typically it performs the opposite action of a primary button

```js

<List>
  <ListItem>
    <Button variant="transparent">Ghost Idle</Button>
  </ListItem>
  <ListItem>
    <Button variant="transparent" state="hover">Ghost Hover</Button>
  </ListItem>
  <ListItem>
     <Button variant="transparent" state="active">Ghost Active</Button>
  </ListItem>
</List>
```

<div class="doc-section-divider"></div>

### Destructive

Destructive Buttons are to be used in situations where you need to convey some very important, potentially irreversible consequence of pressing this button.

```js
<List>
  <ListItem>
    <DestructiveButton>Destructive Button</DestructiveButton>
  </ListItem>
  <ListItem>
    <DestructiveButton state="hover">Destructive Hover</DestructiveButton>
  </ListItem>
  <ListItem>
     <DestructiveButton state="active">Destructive Active</DestructiveButton>
  </ListItem>
</List>
```

### Destructive Variants

Destructive Buttons extend the default button, so they also have different variants
```js
<List>
  <ListItem>
    <DestructiveButton>Destructive Button</DestructiveButton>
  </ListItem>
  <ListItem>
    <DestructiveButton variant="outline">Destructive Outline</DestructiveButton>
  </ListItem>
  <ListItem>
     <DestructiveButton variant="transparent">Destructive Transparent</DestructiveButton>
  </ListItem>
</List>
```

## Size

Use the size property on a `<Button />` to modify the size the button rendered. You can combine it with the `mode` property to get the correct style and size of button you need.

```js
<List>
  <ListItem><Button size="large">Large Button</Button></ListItem>
  <ListItem><Button>Default Button</Button></ListItem>
  <ListItem><Button size="small">Small Light Button</Button></ListItem>
  <ListItem><Button size="xsmall">Extra-Small Button</Button></ListItem>
</List>
```

## Disabled
Use a disabled button to indicate to the user what action will be possible on a page once a prerequisite action is taken. Disabled buttons do not respond to user interaction.
```js
<Button disabled>I am disabled</Button>
```

## Additional Action Considerations
When a call to action requires additional steps, include an ellipsis at the end of the button text as a visual clue. This helps reinforce that there are additional steps that can take place before the action takes effect.
```js
<List>
  <ListItem><Button mode="scary">Move to Trash...</Button></ListItem>
  <ListItem><Button>Print...</Button></ListItem>
</List>
```
