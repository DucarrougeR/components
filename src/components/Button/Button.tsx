import * as React from 'react'
import styled, { StyledComponentClass, css } from '../../styled_components'
import { rem, rgba } from 'polished'
import { brandFont } from '../../styles/typography'
import { ThemeInterface } from '../../themes'
import { themeColors } from '../../themes/theme_colors';

export enum ButtonSizes {
  ExtraSmall = "xsmall",
  Small = "small",
  Medium = "medium",
  Large = "large"
}

export enum ButtonVariants {
  Outline = 'outline',
  Transparent = 'transparent'
}

export interface ButtonProps {
  size: ButtonSizes
  variant: ButtonVariants
}

function buttonSize(size: string) {
  switch (size) {
    case ButtonSizes.ExtraSmall:
      return {
        fontSize: rem(12),
        lineHeight: rem(20),
        padding: `${rem(2)} ${rem(8)}`
      }
    case ButtonSizes.Small:
      return {
        fontSize: rem(14),
        lineHeight: rem(24),
        padding: `${rem(3)} ${rem(14)}`
      }
    case ButtonSizes.Large:
      return {
        fontSize: rem(22),
        lineHeight: rem(30),
        padding: `${rem(8)} ${rem(24)}`
      }
    case ButtonSizes.Medium:
    default:
      return {
        fontSize: rem(16),
        lineHeight: rem(24),
        padding: `${rem(6)} ${rem(16)}`
      }
  }
}

export const Button = styled<ButtonProps, 'button'>('button')`
  --primary: ${props => props.theme.colors.primary };
  --hover: ${props => props.theme.colors.primaryDark };
  --active: ${props => props.theme.colors.primaryDarker };
  --transparentActive: ${props => props.theme.colors.primaryLighter };
  --variantBorder: ${props => props.theme.colors.borderColor };
  --textColor: ${props => props.theme.colors.primaryText };
  --accessibilityOutline ${props => rgba(props.theme.colors.primary, .25)};



  background-color: var(--primary);
  border: ${rem(1)} solid var(--primary);
  border-radius: ${rem(4)};
  color: var(--textColor);
  cursor: pointer;
  display: inline-flex;
  font-family: ${brandFont};
  font-size: ${props => buttonSize(props.size).fontSize};
  line-height: ${props => buttonSize(props.size).lineHeight};
  outline: none;
  padding: ${props => buttonSize(props.size).padding};
  transition: border 80ms,
  vertical-align: middle;
  white-space: nowrap;

  &:hover, &:focus, &.hover {
    border-color: var(--hover);
    background-color: var(--hover);
  }

  &:focus {
    box-shadow: 0 0 0 .15em var(--accessibilityOutline);
  }

  &:active, &.active  {
    border-color: var(--active);
    background-color: var(--active);
    box-shadow: none;
  }

  &[disabled] {
    filter: grayscale(.3);
    opacity: .25;
    cursor: default;

    &:hover, &:active, &:focus {
      background-color: var(--primary);
    }
  }

  ${props => (props.variant === 'outline') && css`
    background: #fff;
    border-color: var(--variantBorder);
    color: var(--primary);

    &:hover, &:focus, &.hover {
      background: #fff;
      color: var(--primary);
      border-color: var(--primary)
    }

    &:active, &.active {
      background: var(--primary);
      color: var(--textColor);
    }
  `};

  ${props => (props.variant === 'transparent') && css`
    background: transparent;
    border: transparent;
    color: var(--primary);

    &:hover, &:focus, &.hover {
      color: var(--hover);
      background: transparent;
    }

    &:active, &.active {
      background: var(--transparentActive);
    }
  `};
`
