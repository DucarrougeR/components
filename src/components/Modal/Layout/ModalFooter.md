`ModalFooter` is attached to the bottom of modal. Generally, the terminal action for a modal as well as an option to abandon the process done via the modal are available within this component. This component is generally used in combination with the `ModalContext.Provider` to get access to the `ModalContext.closeModal()` method.

ModalFooter supports all [`Box`](/#!/Box) properties.

```js
<ModalFooter>
  <Button mx="medium" variant="transparent">
    Cancel
  </Button>
  <Button>Do stuff...</Button>
</ModalFooter>
```
