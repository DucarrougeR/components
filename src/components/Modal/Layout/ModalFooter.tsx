import * as React from 'react'
import { styled } from '../../../style'
import { Box, BoxProps } from '../../Box'

export interface ModalFooterProps extends BoxProps<HTMLDivElement> {
  /**
   * Content that will be placed inside the DialogHeader
   * @required
   */
  children: React.ReactNode

  /**
   * Secondary content to place in the footer
   */
  secondary?: React.ReactNode
}

export const ModalFooter: React.FC<ModalFooterProps> = ({
  children,
  secondary,
  ...props
}) => {
  /*
   * @TODO / Note: When chrome supports `flex-basis: fit-content` minHeight can be removed
   */
  return (
    <Layout
      alignItems="center"
      display="flex"
      is="footer"
      flexDirection="row-reverse"
      {...props}
    >
      <SpaceChildren flexDirection="row-reverse">{children}</SpaceChildren>
      {secondary && <SpaceChildren mr="auto">{secondary}</SpaceChildren>}
    </Layout>
  )
}

// Subtract margin placed on children from the horizonal padding applied to the component
// so that the content is properly aligned.

// padding is: vertical: large, horizonal: calc(xlarge - xsmall)
const Layout = styled(Box)`
  padding: ${props => props.theme.space.large}
    calc(${p => p.theme.space.xlarge} - ${p => p.theme.space.xsmall});
`

const SpaceChildren = styled(Box)`
  display: flex;

  & > * {
    margin-right: ${props => props.theme.space.xsmall};
    margin-left: ${props => props.theme.space.xsmall};
  }
`
