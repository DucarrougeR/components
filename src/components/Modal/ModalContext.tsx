import * as React from 'react'

export interface ModalContextProps {
  closeModal?: (event?: React.SyntheticEvent, doCallbacks?: boolean) => void
}

const context: ModalContextProps = {}

export const ModalContext = React.createContext(context)

export function withModal<T extends ModalContextProps>(
  Component: React.ComponentType<T>
) {
  return function wrapContext(props: T) {
    return (
      <ModalContext.Consumer>
        {overlay => <Component {...props} closeModal={overlay.closeModal} />}
      </ModalContext.Consumer>
    )
  }
}
