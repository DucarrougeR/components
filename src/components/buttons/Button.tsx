import * as React from 'react'
import * as classNames from 'classnames'
import * as styles from './buttons.scss'
import { Button as SButton } from 'semantic-ui-react'

interface ButtonProps {
  className?: string
  danger?: boolean
  onClick?: (...args: any[]) => void
}

export const Button: React.StatelessComponent<ButtonProps> = ({children, className, danger, onClick}) => {
  const btnClass = classNames(
    styles.btn,
    className,
    {
      [styles.btnDanger]: danger
    })

  return <SButton className={btnClass} onClick={onClick ? onClick : () => {}}>{children}</SButton>
}

