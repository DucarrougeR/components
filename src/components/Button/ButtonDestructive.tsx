import * as React from 'react'
import styled, { StyledComponentClass } from '../../styled_components'
import { ThemeInterface } from '../../themes'
import { Button, ButtonProps } from './Button'

export const ButtonDestructive = Button.extend`
  --background-color: ${props => props.variant? 'transparent' : props.theme.colors.destructive };
  --color: ${props => props.variant? props.theme.colors.destructive : props.theme.colors.destructiveText };
  --border-color: ${props => props.variant == 'transparent'? 'transparnet' : props.theme.colors.destructive };
  --hover-color: ${props => props.variant? props.theme.colors.destructiveLightOver : props.theme.colors.destructiveOver };
  --active-color: ${props => props.variant? props.theme.colors.destructiveLightPress : props.theme.colors.destructivePress };
`
