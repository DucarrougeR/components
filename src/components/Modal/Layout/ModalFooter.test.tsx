import 'jest-styled-components'
import * as React from 'react'
import { assertSnapshot } from '../../../../test/utils/snapshot'

import { Button } from '../../Button'
import { ModalContext } from '../ModalContext'
import { ModalFooter } from './ModalFooter'

test('ModalFooter with Button', () => {
  assertSnapshot(
    <ModalFooter>
      <Button variant="transparent" mr="xsmall">
        Cancel
      </Button>
    </ModalFooter>
  )
})

/* tslint:disable:jsx-no-lambda */

test('ModalFooter with ModalContext', () => {
  assertSnapshot(
    <ModalContext.Consumer>
      {({ close }) => (
        <ModalFooter>
          <Button onClick={close} variant="transparent" mr="small">
            Cancel
          </Button>
          <Button
            onClick={() => {
              alert("doin' things...")
              close && close()
            }}
            color="danger"
          >
            Yes, Delete "Stuff"
          </Button>
        </ModalFooter>
      )}
    </ModalContext.Consumer>
  )
})

/* tslint:enable:jsx-no-lambda */
