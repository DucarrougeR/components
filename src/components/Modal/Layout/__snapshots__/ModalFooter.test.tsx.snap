// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalFooter with Button 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.c3 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.c5 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0.25rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: 'Open Sans',Helvetica,Arial,sans-serif;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
}

.c4 {
  font-weight: 600;
  outline: none;
  -webkit-transition: border 80ms;
  transition: border 80ms;
  vertical-align: middle;
  white-space: nowrap;
  border-style: solid;
  border-width: 0.0625rem;
  background: transparent;
  border-color: transparent;
  color: #64518A;
}

.c4:focus {
  box-shadow: 0 0 0 0.15rem rgba(100,81,138,0.25);
}

.c4[disabled] {
  cursor: default;
  -webkit-filter: grayscale(0.3);
  filter: grayscale(0.3);
  opacity: 0.25;
}

.c4[disabled]:hover,
.c4[disabled]:active,
.c4[disabled]:focus {
  background-color: transparent;
  border-color: transparent;
  color: #64518A;
}

.c4:hover,
.c4:focus,
.c4.hover {
  background: #F3F2FF;
  border-color: #F3F2FF;
  color: #6C43E0;
}

.c4:active,
.c4.active {
  background: #E8E5FF;
  border-color: #E8E5FF;
  color: #6C43E0;
}

.c0 {
  padding: 1.25rem calc(2rem - 0.5rem);
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 > * {
  margin-right: 0.5rem;
  margin-left: 0.5rem;
}

<footer
  className="c0 c1"
>
  <div
    className="c2 c3"
  >
    <button
      className="c4 c5"
      style={
        Object {
          "lineHeight": "34px",
        }
      }
    >
      Cancel
    </button>
  </div>
</footer>
`;

exports[`ModalFooter with ModalContext 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.c3 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.c5 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0.25rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: 'Open Sans',Helvetica,Arial,sans-serif;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
}

.c7 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  cursor: pointer;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0.25rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: 'Open Sans',Helvetica,Arial,sans-serif;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
}

.c4 {
  font-weight: 600;
  outline: none;
  -webkit-transition: border 80ms;
  transition: border 80ms;
  vertical-align: middle;
  white-space: nowrap;
  border-style: solid;
  border-width: 0.0625rem;
  background: transparent;
  border-color: transparent;
  color: #64518A;
}

.c4:focus {
  box-shadow: 0 0 0 0.15rem rgba(100,81,138,0.25);
}

.c4[disabled] {
  cursor: default;
  -webkit-filter: grayscale(0.3);
  filter: grayscale(0.3);
  opacity: 0.25;
}

.c4[disabled]:hover,
.c4[disabled]:active,
.c4[disabled]:focus {
  background-color: transparent;
  border-color: transparent;
  color: #64518A;
}

.c4:hover,
.c4:focus,
.c4.hover {
  background: #F3F2FF;
  border-color: #F3F2FF;
  color: #6C43E0;
}

.c4:active,
.c4.active {
  background: #E8E5FF;
  border-color: #E8E5FF;
  color: #6C43E0;
}

.c6 {
  font-weight: 600;
  outline: none;
  -webkit-transition: border 80ms;
  transition: border 80ms;
  vertical-align: middle;
  white-space: nowrap;
  border-style: solid;
  border-width: 0.0625rem;
  background: #CC1F36;
  border-color: #CC1F36;
  color: #FFFFFF;
}

.c6:focus {
  box-shadow: 0 0 0 0.15rem rgba(204,31,54,0.25);
}

.c6[disabled] {
  cursor: default;
  -webkit-filter: grayscale(0.3);
  filter: grayscale(0.3);
  opacity: 0.25;
}

.c6[disabled]:hover,
.c6[disabled]:active,
.c6[disabled]:focus {
  background-color: #CC1F36;
  border-color: #C1C6CC;
}

.c6:hover,
.c6:focus,
.c6.hover {
  background: #B2121F;
  border-color: #B2121F;
}

.c6:active,
.c6.active {
  background: #990F14;
  border-color: #990F14;
}

.c0 {
  padding: 1.25rem calc(2rem - 0.5rem);
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 > * {
  margin-right: 0.5rem;
  margin-left: 0.5rem;
}

<footer
  className="c0 c1"
>
  <div
    className="c2 c3"
  >
    <button
      className="c4 c5"
      style={
        Object {
          "lineHeight": "34px",
        }
      }
    >
      Cancel
    </button>
    <button
      className="c6 c7"
      onClick={[Function]}
      style={
        Object {
          "lineHeight": "34px",
        }
      }
    >
      Yes, Delete "Stuff"
    </button>
  </div>
</footer>
`;
